C/include/c4Certificate.h

```diff
@@ -46,31 +46,27 @@ extern "C" {
-        kC4Cert_CommonName,        // e.g. "Jane Doe", (or "jane.example.com")
-        kC4Cert_Pseudonym,         // e.g. "plainjane837"
-        kC4Cert_GivenName,         // e.g. "Jane"
-        kC4Cert_Surname,           // e.g. "Doe"
-        kC4Cert_Organization,      // e.g. "Example Corp."
-        kC4Cert_OrganizationUnit,  // e.g. "Marketing"
-        kC4Cert_PostalAddress,     // e.g. "123 Example Blvd #2A"
-        kC4Cert_Locality,          // e.g. "Boston"
-        kC4Cert_PostalCode,        // e.g. "02134"
-        kC4Cert_StateOrProvince,   // e.g. "Massachusetts" (or "Quebec", ...)
-        kC4Cert_Country,           // e.g. "us" (2-letter ISO country code)
-        // These are the Subject Alternative Name attributes:
-        kC4Cert_OtherName,         // Binary ASN.1 data containing a type and value
-        kC4Cert_EmailAddress,      // 'rfc822Name', e.g. "jane@example.com"
-        kC4Cert_Hostname,          // 'dnsName', e.g. "www.example.com"
-        kC4Cert_X400Name,          // Yeah, I don't know what this is either
-        kC4Cert_EDIPartyName,      // Ditto
-        kC4Cert_URL,               // "https://example.com/jane"
-        kC4Cert_IPAddress,         // *Binary* IP address, e.g. "\x0A\x00\x01\x01"
-        kC4Cert_RegisteredID,      // Some sort of domain-specific ID?
-
-        kC4Cert_NoAttributeID
-    };
+    typedef C4Slice C4CertNameAttributeID;
+
+    // Some common Distinguished Name attributes:
+    #define kC4Cert_CommonName        C4STR("CN")           // e.g. "Jane Doe", (or "jane.example.com")
+    #define kC4Cert_Pseudonym         C4STR("pseudonym")    // e.g. "plainjane837"
+    #define kC4Cert_GivenName         C4STR("GN")           // e.g. "Jane"
+    #define kC4Cert_Surname           C4STR("SN")           // e.g. "Doe"
+    #define kC4Cert_Organization      C4STR("O")            // e.g. "Example Corp."
+    #define kC4Cert_OrganizationUnit  C4STR("OU")           // e.g. "Marketing"
+    #define kC4Cert_PostalAddress     C4STR("postalAddress")// e.g. "123 Example Blvd #2A"
+    #define kC4Cert_Locality          C4STR("locality")     // e.g. "Boston"
+    #define kC4Cert_PostalCode        C4STR("postalCode")   // e.g. "02134"
+    #define kC4Cert_StateOrProvince   C4STR("ST")           // e.g. "Massachusetts" (or "Quebec", ...)
+    #define kC4Cert_Country           C4STR("C")            // e.g. "us" (2-letter ISO country code)
+
+    // These are the Subject Alternative Name attributes:
+    #define kC4Cert_EmailAddress      C4STR("rfc822Name")   // 'rfc822Name', e.g. "jane@example.com"
+    #define kC4Cert_Hostname          C4STR("dNSName")      // 'dnsName', e.g. "www.example.com"
+    #define kC4Cert_URL               C4STR("uniformResourceIdentifier") // "https://example.com/jane"
+    #define kC4Cert_IPAddress         C4STR("iPAddress")    // *Binary* IP address, e.g. "\x0A\x00\x01\x01"
+    #define kC4Cert_RegisteredID      C4STR("registeredID") // Some sort of domain-specific ID?
```

Comment:
Is it possible to use `constexpr` here instead, or does that not work well with `C4Slice`?  Having so many public preprocessor defines makes me antsy...

Reply:
`constexpr` is C++. In C these would have to be `extern C4String`, and then each of them would be a global symbol that had to be exported from the DLL. :-p